pkg/controller/controller_test.go

```diff
@@ -3156,3 +3129,257 @@ func TestProvisionFromPVC(t *testing.T) {
+				inTreePluginName, false, mockTranslator)
+
+			// Set up return values (AnyTimes to avoid overfitting on implementation)
+			if tc.expectTranslation {
+				// Have fake translation provide same as input SC but with
+				// Parameter indicating it has been translated
+				mockTranslator.EXPECT().TranslateInTreeStorageClassToCSI(gomock.Any(), gomock.Any()).DoAndReturn(
+					func(inTreePluginName string, sc *storagev1.StorageClass) (*storagev1.StorageClass, error) {
+						newSC := sc.DeepCopy()
+						newSC.Parameters[translatedAnnKey] = "foo"
+						return newSC, nil
+					},
+				).AnyTimes()
+				// Have fake translation provide same as input PV but with
+				// Annotation indicating it has been translated
+				mockTranslator.EXPECT().TranslateCSIPVToInTree(gomock.Any()).DoAndReturn(
+					func(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+						newPV := pv.DeepCopy()
+						if newPV.Annotations == nil {
+							newPV.Annotations = map[string]string{}
+						}
+						newPV.Annotations[translatedAnnKey] = "foo"
+						return newPV, nil
+					},
+				).AnyTimes()
+			}
+
+			// Set an expectation that the Create should be called
+			if tc.expectCreate {
+				expectParams := map[string]string{"fstype": "ext3"} // Default
+				if tc.expectTranslation {
+					// If translation is expected we check that the CreateVolume
+					// is called on the expected volume with a translated param
+					expectParams[translatedAnnKey] = "foo"
+				}
+				controllerServer.EXPECT().CreateVolume(gomock.Any(),
+					&csi.CreateVolumeRequest{
+						Name:               "test-testi",
+						Parameters:         expectParams,
+						VolumeCapabilities: nil,
+						CapacityRange: &csi.CapacityRange{
+							RequiredBytes: int64(requestBytes),
+						},
+					}).Return(
+					&csi.CreateVolumeResponse{
+						Volume: &csi.Volume{
+							CapacityBytes: requestBytes,
+							VolumeId:      "test-volume-id",
+						},
+					}, nil).Times(1)
+			}
+
+			// Make a Provision call
+			volOpts := controller.ProvisionOptions{
+				StorageClass: &storagev1.StorageClass{
+					Provisioner:   tc.provisioner,
+					Parameters:    map[string]string{"fstype": "ext3"},
+					ReclaimPolicy: &deletePolicy,
+				},
+				PVName: "test-name",
+				PVC:    createFakePVCWithAnnotation(tc.annotation, requestBytes),
+			}
+
+			pv, state, err := csiProvisioner.(controller.ProvisionerExt).ProvisionExt(volOpts)
+			if tc.expectErr && err == nil {
+				t.Errorf("Expected error, got none")
+			}
+			if err != nil {
+				if !tc.expectErr {
+					t.Errorf("got error: %v", err)
+				}
+				return
+			}
+
+			if controller.ProvisioningFinished != state {
+				t.Errorf("expected ProvisioningState %s, got %s", controller.ProvisioningFinished, state)
+			}
+
+			if tc.expectTranslation {
+				if _, ok := pv.Annotations[translatedAnnKey]; !ok {
+					t.Errorf("got no translated annotation %s on the pv, expected one", translatedAnnKey)
```

Comment:
nit about the error message: What would be most useful in a test failure log? "expected this annotation", or "expected PV to be translated to in-tree"?